rfcs/0000-cosign-integration.md

```diff
@@ -0,0 +1,132 @@
+# Integrate `cosign` with `kpack`
+
+## Problem
+
+The signing solution currently supported by `kpack` is
+[Notary v1](https://github.com/theupdateframework/notary). However, Notary
+presents itself with several different challenges, including:
+- It is unable to sign container image digests. The user must always specify a
+tag to be signed.
+- It is unable to verify container image signature by digests.
+- Signature relocation is not possible with this tool.
+- It is a project that currently has low activity and community involvement,
+  since the community is currently focused on the development of Notary v2.
+
+[`cosign`](https://github.com/sigstore/cosign) presents itself as a solution to
+some of those aforementioned challenges. With `cosign`, the user is able to:
+- Relocate container images across registries along with its signatures.
+  - Currently, there is no built-in command for relocation on `cosign`, but
+    copying container images and signatures across registries using other
+    container CLI tools such as `crane` and then verifying with `cosign` is
+    possible.
+  - There is also a proposal for
+    [implementing a `cosign cp`](https://github.com/sigstore/cosign/issues/303)
+    command that would affect this copy as a single operation.
+- Sign container images by tag or by digest.
+- Verify container image signatures by specifying either a tag or a digest.
+- Verify container image signatures after relocation.
+
+Moreover, `cosign` is a project that has had a lot of activity and community
+involvement recently.
+
+## Outcome
+
+`kpack` integrates with `cosign` to sign images it builds and push the
+signatures to a registry so that users can ensure the chain of custody of a
+generated artifact.
+
+This proposal aims to cover specifically the flow of signing an image produced
+by `kpack Image` resource without verification of base or builder images pulled
+in the process.
+
+## Actions to take
+
+### Enable image signing using `cosign`
+
+- Create a new configuration in the `Image` resource that allow users to
+  configure image signing with `cosign` as the signing mechanism, specifying
+  the secret reference to get the private key and passphrase. The user can also
+  specify signature annotations as part of configuration:
+```yaml
+cosign:
+  # a list of secret references holding the private keys and passphrases for signing images
+  secretRefs:
+  - name: <secret-name 1>
```

Comment:
Could we do without specifying the secrets here and just sign images when they are provided with "cosign" `secrets` on the configured service account? This would allow "cosign" secrets to be configured analogous to the secrets provided for registry and git access. This would make configuring multiple cosigned images per namespace easier?

Reply:
Makes sense. In this case, would we need to have a specific secret name to tell us whether `cosign` signing is being requested?

Reply:
Updated the RFC to state that `kpack` will read the secret types to determine whether `cosign` should be used as the signing mechanism.